### PR TITLE
Feature/logimages

### PR DESCRIPTION
**Pull request by following5**
This no pull request, but a discussion thread for implementing #360. Looks like this is the best way to collect all commit-related information for a feature branch at one place.

### CODE REVIEW

**Comment by following5** [comment]
Are you sure that you want to encourage client developers to work around the OCPL restriction of forcing users to enter image captions? If yes, I think it should be documented that OKAPI may generate dummy captions if no caption is provided. Developers may rather want to force users to enter a caption than producing dummies.

If no, then there is no need to make things that complicated now. A required parameter still can be changed to 'optional' later, if necessary.

**Comment by following5** [comment]
All those OKAPI featues that work different on different OC sites can discourage developers from implementing them at all. :( At least as long there is no api to query all those site capabilities and parameters.

**Comment by wrygiel** [comment]
> Are you sure that you want to encourage client developers to work around the OCPL restriction of forcing users to enter image captions?

What do you mean?

> If yes, I think it should be documented that OKAPI may generate dummy captions if no caption is provided. Developers may rather want to force users to enter a caption than producing dummies.

For now, OKAPI won't generate dummy captions, and it will require a caption to be entered. But this interface gives us the ability to change that in various ways in the future.

> If no, then there is no need to make things that complicated now. A required parameter still can be changed to 'optional' later, if necessary.

But the existing applications won't respect that then (they will still require the caption provided by their user).

**Comment by following5** [comment]
> Looks like this is the best way to collect all commit-related information for a feature branch at one place.

The cleaner way is to include "update #issue" in all commit comments, so they will have a backlink to the issue. But it's easy to forget that ...

**Comment by wrygiel** [comment]
> The cleaner way is to include "update #issue" in all commit comments, so they will have a backlink to the issue. But it's easy to forget that ...

I think pull request is okay in this case. We have a separate branch we're working on, and we are going to merge it once we're done. That's exactly what pull requests are for :)

**Comment by following5** [comment]
> Approximately 20% of these images have useful captions. The rest (80%) do not contain any useful information, just meaningless comments, or random strings.

There is little space for captions in the OCPL log image galleries - not enough for descriptive captions. So the loggers enter garbage.

**Comment by following5** [comment]
AI have used the most common GD library for image processing. But it would be nice to use _Imagick_ instead if available - it would allow to strip EXIF data from JPEGs without resampling the images (which loses some quality).

**Comment by following5** [comment]
With the last commit, I added a table okapi_images, which works like okapi_cache_logs. But now I have realized that there can be many other things submitted through OKAPI, e.g. the geocache reports. Instead of cluttering the database with history tables, it seems more efficient to put that all into one table.

So I will try to change that now. If it's not ok, it can just be reverted.

File: okapi/core.php [hunk 1]
```diff
@@ -1902,6 +1904,26 @@ public static function fix_oc_html($html)
 
         return $html;
     }
+
+    function php_ini_get_bytes($setting)
+    {
+        $setting = ini_get($setting);
+        $setting = str_replace('K', '*1024', $setting);
+        $setting = str_replace('M', '*1024*1024', $setting);
+        $setting = eval('return '.$setting.';');
```
Comment (wrygiel) [inline okapi/core.php]: Are you sure it's safe to use eval in this context? Perhaps `preg_replace("/^[0-9*]+$/", $setting)`?

Comment (wrygiel) [inline okapi/core.php]: (btw, I still haven't read all of your emails, sorry - I decided I will first take a look at the diff)

File: okapi/settings.php [hunk 2]
```diff
@@ -180,6 +185,21 @@
...
+        /**
+         * Maximum 'resolution' of saved images in pixels
+         */
+        'IMAGE_MAX_PIXEL_COUNT' => 524288,  # 0.5 MP
+
+        /**
+         * Quality of saved JPEG images on a scale from 50 to 100. The higher the
+         * quality, the sharper the image and larger the saved image files
+         * (file size grows exponentially). The PHP default is 75, which is also
+         * used in OCPL code. As OKAPI always resamples uploaded images - which
```
Comment (wrygiel) [inline okapi/settings.php]: OKAPI always reasamples - why? :) Probably there's a thread for this, if so, I will find it in a minute.

**Comment by wrygiel** [comment]
> Instead of cluttering the database with history tables, it seems more efficient to put that all into one table.

Ok. I'm fine with both approaches for now.

**Comment by wrygiel** [comment]
I won't have time to review all the code now, but I had a look at all the changes to the interface.

The only thing I'm not entirely comfortable about is lack of the possibility to enforce max upload sizes. It's like saying "you can always push anything you want at us, regardless of its size, and we're required to always support this, indefinitely".

**Comment by wrygiel** [comment]
If we keep it like that, then - in theory - we should support all POST sizes. The external programmer has no way of knowing what the limits of our server are.

**Comment by following5** [comment]
> OKAPI always reasamples - why? :) Probably there's a thread for this, if so, I will find it in a minute.

You fill find it in several places, e.g. the comment above on GD and EXIF. To avoid resamling JPEG for images below pixel limit, one of the following would need to be done:

* Keep EXIF information. This e.g. can lead to accidential spoilering of final coordinates for multies and mysteries, because many modern cameras record GPS location; or to unexpected disclosure of private data,
* use Imagick instead of GD, which currently is only installed at OC.de,
* require some other library for stripping EXIF, or a binary tool where the data is piped through.

**Comment by following5** [comment]
> If we keep it like that, then - in theory - we should support all POST sizes. The external programmer has no way of knowing what the limits of our server are.

There are limits on everything. E.g. the maximum size of a log comment, or the maximum amount of additional OKAPI requests the server can handle just now. And of course the maximum pixel dimensions of an image. But currently none of those limits are exposed.

But if you can implement such a limit in some meaningful way, I won't mind.

**Comment by following5** [comment]
> The only thing I'm not entirely comfortable about is lack of the possibility to enforce max upload sizes. 

_How_ do you want to enforce such a limit, technically?

**Comment by following5** [comment]
Note: All OC sites currently resample all uploaded images and strip EXIF data. OCPL resamples with quality 75, OCDE with quality 85 (see the JPEG_QUALITY setting in settings.php).

**Comment by following5** [comment]
> If we keep it like that, then - in theory - we should support all POST sizes. The external programmer has no way of knowing what the limits of our server are.

The problem is - I don't know a way to know the limits either. I set PHP `post_max_size` to 1 MB, verified that the setting was reflected by `phpinfo()`, and still could send a complete 1.6 MB base64-encoded image to OKAPI.

Maybe you have more luck in implementing this. :)

File: okapi/core.php (hunk 1 above)
Comment (following5) [inline okapi/core.php]: It is safe as long as noone with root rights enters something evil in php.ini. If someone does that, we probably have larger problems than this :)  but I like regex and have inserted one which verifies the setting. This also protects from requesting a non-numeric variable here.

The alternative to eval() would be extracting the number and (optional) multiplier and do the maths in direct PHP code.

Comment (wrygiel) [inline okapi/core.php]: Oh, I didn't notice that it's just from php.ini, sorry. I thought that you parse the value passed in `$setting`.

**Comment by following5** [comment]
I am not sure if you noticed this: I did a change to the definition of logs/images/edit, requiring that at least one of the optional parameters is given. The main reason is that I did not know what `success` message to return in this case: "Your image has successfuly not been updated"?

The overall implementation is complete, and I think I have tested everything on both OCDE and OCPL code.

**Comment by wrygiel** [comment]
> The main reason is that I did not know what success message to return in this case: "Your image has successfuly not been updated"?

The most probably scenario of this occurring would be that of a user clicking "edit" and then clicking "save" on some form without editing the form. I guess that in this case the message "successfully updated" is still okay.

I will have a last look and confirm there's nothing wrong with the interface.

If you want, I can also try to review the code, but this would take longer.

**Comment by wrygiel** [comment]
Done. Made some descriptions more explicit. Check them out to make sure if I understand them the way you wanted them understood. If yes, then feel free to merge.

**Comment by following5** [comment]
I have opened an [OCPL pull request](https://github.com/opencaching/opencaching-pl/pull/430) for updating okapi_settings.php.

**Pull request merged**